codec-http/src/main/java/io/netty/handler/codec/http/multipart/HttpPostMultipartRequestDecoder.java

```diff
@@ -1035,7 +1035,7 @@ private static String readLine(ByteBuf undecodedChunk, Charset charset) {
         }
         SeekAheadOptimize sao = new SeekAheadOptimize(undecodedChunk);
         int readerIndex = undecodedChunk.readerIndex();
-        ByteBuf line = undecodedChunk.alloc().heapBuffer(64);
+        ByteBuf line = undecodedChunk.alloc().heapBuffer(64, 64);
```

Comment:
@fredericBregier can you explain why setting an upper limit improves things here ?

Reply:
@normanmaurer 
I don't have any real clue, but it seems it does. That's why I'm asking you to look at this why. 

The change is huge on performances in PARANOID mode (using the example test given by the final user), about 10 times better... I test several times, and without it, the times are about 1000ms, while with it, it gives about 100ms.

Using my old test (a bit different, where a lot of items, about 6000, are passed and not only one file), the improvement is there also on PARANOID mode, but not that much (not 10 times better, comparing 146.000 old code vs 105.000 new code). 
There, the issue (and change) is on the `undecodedChunk.write(buf)`.
But if I change it to something like `wrappedBuffers(undecodedChunk, buf)`, the time is close to 1000ms again in PARANOID mode, but increases sensibly in other modes (500ms against 200ms without any change).


It was strange for me to see that comparing:
- `undecodedChunk = isLast? ... : buf.alloc.buffer(buf.readableSize(), buf.readableSize()).write(buf)` and `undecodedChunk = Unpooled.wrapped(undecodedChunk, buf.alloc.buffer(buf.readableSize(), buf.readableSize()).write(buf))`
- vs original ones (still in place at lines 338 and 340

gives impact performances (better for PARANOID, about 1000 times, but worst in other modes, about 2 times). I feel like it was the same in both cases (allocation and writing), but it seems not. Note that in current code, `undecodedChunk` in line 338 is still allocated using no upper bound.

The main "same root cause" seems to be "no upper bound for buffer" when using PARANOID mode, but in different ways.


I feel like this proposal is safe as it enhances a bit the general performances and takes down the PARANOID issue, at least for the user's example.
But I feel also that there is something else, since in my test (high number of items) this is really not enough.

So the reason I try to test it and produce at least this in order to get help.

Reply:
@normanmaurer 

I added a fake commit to let you run the different cases at once:

This commit is only intend to show the result on tests using 3 cases:
- HttpPostMultipartRequestDecoder.TEST_TEMP_ITEM = 1: original case
- HttpPostMultipartRequestDecoder.TEST_TEMP_ITEM = 2: where only upper bound is modified
- HttpPostMultipartRequestDecoder.TEST_TEMP_ITEM = 3: where write is replaced by wrapped and the last "no upper bounded" buffer is replaced too

Test to run is `testRegressionMultipleLevelLeakDetector`.

On my side results are:

Timer: DISABLED NotUsingDisk1 => 220.995585
**Timer: SIMPLE NotUsingDisk1 => 222.134248**
Timer: ADVANCED NotUsingDisk1 => 208.68989
**_Timer: PARANOID NotUsingDisk1 => 100044.293876_**
Timer: DISABLED UsingDisk1 => 707.480873
Timer: DISABLED UsingDisk1 => 518.695771
Timer: SIMPLE UsingDisk1 => 522.790849
Timer: ADVANCED UsingDisk1 => 585.759164
Timer: PARANOID UsingDisk1 => 105090.135276

Timer: DISABLED NotUsingDisk2 => 136.461289
**Timer: SIMPLE NotUsingDisk2 => 126.793194**
Timer: ADVANCED NotUsingDisk2 => 129.771282
**_Timer: PARANOID NotUsingDisk2 => 973.38734_**
Timer: DISABLED UsingDisk2 => 74.279745
Timer: DISABLED UsingDisk2 => 71.332363
Timer: SIMPLE UsingDisk2 => 114.811827
Timer: ADVANCED UsingDisk2 => 89.290387
Timer: PARANOID UsingDisk2 => 1155.278418

Timer: DISABLED NotUsingDisk3 => 1829.364849
**Timer: SIMPLE NotUsingDisk3 => 1949.512746**
Timer: ADVANCED NotUsingDisk3 => 2120.573799
**_Timer: PARANOID NotUsingDisk3 => 2073.506771_**
Timer: DISABLED UsingDisk3 => 1998.837595
Timer: DISABLED UsingDisk3 => 1952.342478
Timer: SIMPLE UsingDisk3 => 2099.951834
Timer: ADVANCED UsingDisk3 => 2058.028121
Timer: PARANOID UsingDisk3 => 2295.847557

Reply:
Last fake commit:

This commit is only intend to show the result on tests using 3 cases:
    - HttpPostMultipartRequestDecoder.TEST_TEMP_ITEM = 1: original case
    - HttpPostMultipartRequestDecoder.TEST_TEMP_ITEM = 2: where only upper bound is modified
    - HttpPostMultipartRequestDecoder.TEST_TEMP_ITEM = 3: where temptative to reuse more the existing ByteBuf is done
    
Test to run is `testRegressionMultipleLevelLeakDetectorNoDisk` or `testHighNumberCheckLeakDetectorVersions`
    
Current results (stability is correct but numbers depend on host)

I've done various tests:
- using `wrapped` (or through `buf.alloc().compositeByteBuf()`): worst timers
- using fixed allocation (but high enough to fit,the necessary elements, just for test): not good timers (probably because the size is too huge for a lot of cases, and whatever, it is not acceptable to have fixed size, it was just for testing purpose)
- using various ways to discard read bytes: it gives a bit improvement (TEST_TEMP_ITEM=3): almost best results, but not that much

In particular, in the "big" test (testHighNumberCheckLeakDetectorVersions where more than 16000 items are sent, while the other test is only one item from a big file), I am not able to reduce the timer in PARANOID mode, while in "unique" item test ( testRegressionMultipleLevelLeakDetectorNoDisk), the PARANOID decreases a lot.

Even if I cannot understand why setting the upper bound brings such a better improvement, I feel like this is at least a good change.
However, for the other try using discardReadBytes (or tries using wrapped buffers), it seems to not enhanced the timers.

@normanmaurer If you have any clue to continue, I will. I would propose to at least have the minimal change (upper bounds) in order to fix a bit this issue in PARANOID mode (and does not changed a lot, better or worst, for other levels).

_Here are my results (I tried to use profiling to understand the reasons, but the profiling just kill all results, therefore I was unable to deep in buffer implementations to see where is the main difference)._


testHighNumberCheckLeakDetectorVersions
=======================================

HighItemNumberDISABLED1=506.43838600000004, 
HighItemNumberDISABLED2=496.729877, 
HighItemNumberDISABLED3=503.72193200000004, 

**_HighItemNumberSIMPLE1=460.46265600000004, 
HighItemNumberSIMPLE2=475.950721, 
HighItemNumberSIMPLE3=467.69606699999997,_** 

HighItemNumberADVANCED1=465.508472, 
HighItemNumberADVANCED2=611.797092, 
HighItemNumberADVANCED3=467.812221, 

**HighItemNumberPARANOID1=304098.68770899996, 
HighItemNumberPARANOID2=304140.256149, 
HighItemNumberPARANOID3=303301.14581200003,** 

BigItemDISABLED1=426.205971, 
BigItemDISABLED2=441.343974, 
BigItemDISABLED3=420.528978, 

**_BigItemSIMPLE1=420.825118, 
BigItemSIMPLE2=432.968082, 
BigItemSIMPLE3=418.955781,_** 

BigItemADVANCED1=420.391646, 
BigItemADVANCED2=445.089119, 
BigItemADVANCED3=428.97732599999995, 

**BigItemPARANOID1=318339.78715600003
BigItemPARANOID2=319121.713933
BigItemPARANOID3=316866.262051**

testRegressionMultipleLevelLeakDetectorNoDisk
=============================================

Timer: DISABLED NotUsingDisk1 => 195.185296
Timer: DISABLED NotUsingDisk2 => 206.28149
Timer: DISABLED NotUsingDisk3 => 142.354934

**_Timer: SIMPLE NotUsingDisk1 => 185.459894
Timer: SIMPLE NotUsingDisk2 => 82.243331
Timer: SIMPLE NotUsingDisk3 => 83.65489_**

Timer: ADVANCED NotUsingDisk1 => 175.245971
Timer: ADVANCED NotUsingDisk2 => 214.114222
Timer: ADVANCED NotUsingDisk3 => 89.005829

**Timer: PARANOID NotUsingDisk1 => 119286.301819
Timer: PARANOID NotUsingDisk2 => 1604.135822
Timer: PARANOID NotUsingDisk3 => 2489.555437**